crates/sui-indexer-alt/src/handlers/mod.rs

```diff
@@ -0,0 +1,634 @@
+
+        // Track the high watermark for the pipeline. The pipeline confirms that it has written all
+        // checkpoint data up from the watermark it is initialised with up to and including this
+        // watermark.
+        //
+        // To correctly update the watermark, the committer tracks the watermark it last tried to
+        // write and the watermarks for any checkpoints that have been written since then
+        // ("pre-committed"). After each batch is written, the committer will try to progress the
+        // watermark as much as possible without going over any holes in the sequence of
+        // checkpoints.
+        //
+        // NOTE: When no watermark is provided, it is assumed that the pipeline is starting from
+        // scratch, but we still initialize it as if it is at (after) the genesis checkpoint. This
+        // means we never write a watermark for the genesis checkpoint, but would wait for another
+        // checkpoint to be written out before updating the watermark, which is fine in practice
+        // and simplifies the logic of tracking watermarks.
+        let mut precommitted: BTreeSet<CommitterWatermark<'static>> = BTreeSet::new();
+        let mut watermark =
+            watermark.unwrap_or_else(|| CommitterWatermark::initial(H::NAME.into()));
+
+        // The committer will periodically output a log message at a higher log level to
+        // demonstrate that the pipeline is making progress.
+        let mut next_loud_watermark_update =
+            watermark.checkpoint_hi_inclusive + LOUD_WATERMARK_UPDATE_INTERVAL;
+
+        info!(pipeline = H::NAME, ?watermark, "Starting committer");
+
+        loop {
+            tokio::select! {
+                // Break ties in favour of operations that reduce the size of the buffer.
+                biased;
+
+                _ = cancel.cancelled() => {
+                    info!(pipeline = H::NAME, "Shutdown received, stopping committer");
+                    break;
+                }
+
+                // Time to write out another batch of rows, if there are any.
+                _ = poll.tick(), if pending_rows > 0 => {
+                    let Ok(mut conn) = db.connect().await else {
+                        warn!(pipeline = H::NAME, "Failed to get connection for DB");
+                        cool.reset();
+                        continue;
+                    };
+
+                    let guard = metrics
+                        .committer_gather_latency
+                        .with_label_values(&[H::NAME])
+                        .start_timer();
+
+                    while batch_values.len() < H::CHUNK_SIZE {
+                        let Some(mut entry) = pending.first_entry() else {
+                            break;
+                        };
+
+                        let indexed = entry.get_mut();
+                        let values = &mut indexed.values;
+                        if batch_values.len() + values.len() > H::CHUNK_SIZE {
```

Comment:
Let's leave `objects_snapshot` aside, because the way I would solve that is by adding an explicit buffer task on an existing `objects` pipeline, and partitioning, because all the tables this pipeline implementation caters to are either KV tables which won't be in postgres in the long run or tables that aren't currently dynamically partitioned, but tell me more about the requirement related to the Governance API, because I don't have enough context to answer that, but it may be an interesting exercise to see how it fits into this architecture.

Reply:
okay we can leave partition and objects_snapshot aside for now, @emmazzz prob knows more about Governance API.